### PR TITLE
feat(Preferences): Add std::string to the Preferences object.

### PR DESCRIPTION
**Pull request by nielsnl68**
*By completing this PR sufficiently, you help us to review this Pull Request quicker and also help improve the quality of Release Notes*

### Checklist
1. [x] Please provide specific title of the PR describing the change, including the component name (*eg. „Update of Documentation link on Readme.md“*)
2. [x] Please provide related links (*eg. Issue which will be closed by this Pull Request*)
3. [x] Please **update relevant Documentation** if applicable
4. [x] Please check [Contributing guide](https://docs.espressif.com/projects/arduino-esp32/en/latest/contributing.html)
5. [x] Please **confirm option to "Allow edits and access to secrets by maintainers"** when opening a Pull Request

*This entire section above can be deleted if all items are checked.*

-----------
## Description of Change
In my firmware i use only the std::string as string storage's.
When i want to use the Preferences object i found out that only the Arduino String variable is offered.
This PR add this option. It is basically a copy of the `putString()` and `getString()` but returning `std::string` instead.

## Test Scenarios
Please describe on what Hardware and Software combinations you have tested this Pull Request and how.
I have tested my Pull Request on Arduino-esp32 core v3.3.10 with an ESP32-S3.
(*eg. I have tested my Pull Request on Arduino-esp32 core v2.0.2 with ESP32 and ESP32-S2 Board with this scenario*)
```cpp
std::string current_name = "";

void load() {
  Preferences prefs;
  prefs.begin("testing", true);
  current_name = prefs.getStdString("name", current_name);
  prefs.end();
}

void load() {
  Preferences prefs;
  prefs.begin("testing");
  prefs.putStdString("name", current_name);
  prefs.end();
}
```

## Related links
Please provide links to related issue, PRs etc.
https://github.com/espressif/arduino-esp32/issues/12831
https://github.com/espressif/arduino-esp32/discussions/12820
(*eg. Closes #number of issue*)

### CODE REVIEW

File: docs/en/tutorials/preferences.rst
[the remark this answers is not in the document]
Comment (nielsnl68) [reply docs/en/tutorials/preferences.rst:1]: link does not work properly.

**Comment by nielsnl68** [comment]
Sorry, Altrough i understand why this PR fails, i do not know how resolve the issue correctly. 
When someone likes to jump in then i'm really happy,

File: libraries/Preferences/src/Preferences.h [hunk 1]
```diff
@@ -16,6 +16,10 @@
 
 #include "Arduino.h"
 
+#if __has_include(<string>)
```
Comment (robertlipe) [inline libraries/Preferences/src/Preferences.h:19]: I'm not sure you need this test. In all but the most restrictive environments (freestanding, which is already doomed with Arduino) you're going to have <string> - that's been part of the STL since the early 90's.

File: libraries/Preferences/src/Preferences.cpp [hunk 2]
```diff
@@ -516,6 +522,29 @@ String Preferences::getString(const char *key, const String defaultValue) {
   return String(buf);
 }
 
+#ifdef USE_STD_STRING
+std::string Preferences::getStdString(const char *key, const std::string defaultValue) {
```
Comment (robertlipe) [inline libraries/Preferences/src/Preferences.cpp:526]: Make that a string& and you'll save a copy on every call.

File: libraries/Preferences/src/Preferences.cpp [hunk 3]
```diff
@@ -516,6 +522,29 @@ String Preferences::getString(const char *key, const String defaultValue) {
...
+  if (!_started || !key) {
+    return std::string(defaultValue);
+  }
+  esp_err_t err = nvs_get_str(_handle, key, value, &len);
+  if (err) {
+    log_e("nvs_get_str len fail: %s %s", key, nvs_error(err));
+    return std::string(defaultValue);
+  }
+  char buf[len];
+  value = buf;
```
Comment (robertlipe) [inline libraries/Preferences/src/Preferences.cpp:538]: IMO, it's clearer not reuse value here. Just pass &buf.

File: libraries/Preferences/src/Preferences.cpp [hunk 4]
```diff
@@ -516,6 +522,29 @@ String Preferences::getString(const char *key, const String defaultValue) {
...
+  err = nvs_get_str(_handle, key, value, &len);
+  if (err) {
+    log_e("nvs_get_str fail: %s %s", key, nvs_error(err));
+    return std::tring(defaultValue);
```
Comment (robertlipe) [inline libraries/Preferences/src/Preferences.cpp:542]: How does that compile?

I'm not saying it's a great idea, but you might be able to save one trip through nvs_get_str just by deciding that MOST keys will be shorter than some reasonable value and at least trying to shove it into that buffer, just reallocating on failure.

const int kDefaultSize  = 65;
char* buf = alloca(kDefaultSize); // Why alloca? so you can reallocate it.
// try to get it. If len didn't come back 65, return std::string(buf); and you're done.
Now you have the real size of len you can use for nvs_get_str()

It's a real annoyance to bolt C++ APIs atop C for this kind of thing.

Maybe instead of alloca, you can malloc it and then std::move the result into a string.

It's definitely the kind of thing I would wrap up in a unit test that exercised BOTH paths of that with values below, at, and above the knee.

I wouldn't fake shock if the C API had edge cases involving null termination. Test it with a long key to see.           

While I'm all about abolishing String from the world, if this is an Arduino library and not a 'normal' C++ one, should we accept the self-loathing and make that a String instead of a std::string? (Insert "Why not both?" meme)

**Review by robertlipe — COMMENTED** [review]
The hint you probably need is buried inside the error itself.


/__w/arduino-esp32/arduino-esp32/components/arduino-esp32/libraries/Preferences/src/Preferences.h:69:40: error: 'std::string' has not been declared
   69 |   size_t putStdString(const char *key, std::string value);
      |                                        ^~~
/__w/arduino-esp32/arduino-esp32/components/arduino-esp32/libraries/Preferences/src/Preferences.h:91:8: error: 'string' in namespace 'std' does not name a type
   91 |   std::string getStdString(const char *key, std::string defaultValue = std::string());
      |        ^~~~~~
/__w/arduino-esp32/arduino-esp32/components/arduino-esp32/libraries/Preferences/src/Preferences.h:18:1: note: 'std::string' is defined in header '<string>'; this is probably fixable by adding '#include <string>'
   17 | #include "Arduino.h"
  +++ |+#include <string>

You're testing for the presence of string (why? it's been part of the STL since...forever) but never actually including it.

File: libraries/Preferences/src/Preferences.cpp (hunk 4 above)
Comment (robertlipe) [reply libraries/Preferences/src/Preferences.cpp:542]: Another idea. Eliminate the special case. The code you need is in the very next function.

need_len = getStringLength(key);
buf[need_len + 1];
  esp_err_t err = nvs_get_str(_handle, key, &buf, &len); // Again, consider zero terminatein rules
return std::string(buf)

Done. This works because the next function IS the "call it with zero just so it can tell me the size" case. 

This means you're still making two nvs_get_str API calls, but the code becomes dead easy. (Well, other than the ambiguity on whether 'len' includes the null terminator or not.)

Comment (nielsnl68) [reply libraries/Preferences/src/Preferences.cpp:542]: Nice idea. But that is not fully the issue here.
It seams that **std::string** is not included by default. I do not want to include the `<string>` header by default just to allow nice need.

Comment (robertlipe) [reply libraries/Preferences/src/Preferences.cpp:542]: It's not optional. That's what the error is telling you quite clearly. In C++, you have to include what you use; hoping that someone else included it for you rarely pays off.

And the "have a small buffer, if it doesn't fit, make a bigger one" is in code that I *just* had to write and test. It's algorthimically perfectly to drop in here if you choose to go that way.

```
  static string asprintf(const std::string& fmt, ...) {
+    char buf[1024];
+    va_list args;
+    va_start(args, fmt);
+    int n = vsnprintf(buf, sizeof(buf), fmt.c_str(), args);
+    va_end(args);
+    if (n < 0) {
+      return string();
+    }
+    if (n < static_cast<int>(sizeof(buf))) {
+      return string(buf);
+    }
+    string result(n, '\0');
+    va_start(args, fmt);
+    vsnprintf(result.data(), n + 1, fmt.c_str(), args);
+    va_end(args);
+    return result;
+  }
```
